File: String/validAnagram.ts
/**
 * LeetCode #242 - Valid Anagram
 * Link: https://leetcode.com/problems/valid-anagram/
 *
 * Problem:
 * Given two strings s and t, return true if t is an anagram of s, 
 * and false otherwise.
 *
 * @time  O(n) - Where n is the length of the strings.
 * @space O(1) - Constant space if restricted to 26 lowercase letters, 
 * or O(k) for Unicode characters.
 */
function isAnagram(s: string, t: string): boolean {
   if (s.length !== t.length) {
    return false;
  }

  const charCount: Record<string, number> = {};

  for (const char of s) {
    charCount[char] = (charCount[char] || 0) + 1;
  }

  for (const char of t) {
    if (!charCount[char]) {
      return false;
    }

    charCount[char]--;
  }

  return true;
};

File: Strings/valid_anagram.py
'''
Check Anagram (Techie Delight #40)
Links:https://www.techiedelight.com/?problem=CheckAnagram

Problem:
Given two strings X and Y, determine if they are anagrams of each other.
An anagram is a word formed by rearranging the letters of another.

Example:
Input: X = 'silent', Y = 'listen'
Output: True

Complexity Analysis:
@time  O(n) - We iterate through the strings to build a frequency map.
@space O(k) - Where k is the number of unique characters.
'''

class Solution:
  def isAnagram(self, X: str, Y: str) -> bool:
       if len(X) != len(Y):
           return False

       char_counts = {}

       for char in X:
           char_counts[char] = char_counts.get(char, 0) + 1

       for char in Y:
           if char not in char_counts:
               return False
            
           char_counts[char] -= 1
            
           if char_counts[char] < 0:
               return False

       return all(count == 0 for count in char_counts.values())
